File: packages/cli/src/cli/http.ts
import { buildCommand } from "@stricli/core";
import type { ProxerCommandContext } from "#app/application.ts";
import { preferFlag, readEnvString } from "#app/cli/env.ts";
import { runHttpClient } from "#app/cli/run.ts";
import {
  basicAuthPasswordFlag,
  basicAuthUsernameFlag,
  httpServerFlag,
  httpSubdomainFlag,
  parseHttpSubdomain,
  tokenFlag,
} from "#app/cli/shared-flags.ts";
import { DEFAULT_HTTP_SERVER_URL } from "#app/config/constants.ts";
import { resolveControlServerUrl } from "#app/lib/control-url.ts";
import { ProxerError } from "#app/lib/error.ts";

type HttpFlags = {
  readonly basicAuthPassword?: string;
  readonly basicAuthUsername?: string;
  readonly server?: string;
  readonly subdomain?: string;
  readonly token?: string;
};

type HttpRouteRequest =
  | { readonly type: "auto" }
  | { readonly type: "root" }
  | { readonly type: "subdomain"; readonly subdomain: string };

const parseLocalPort = (input: string): number => {
  if (!/^\d+$/.test(input)) {
    throw new ProxerError("local port must be a number");
  }

  const port = Number(input);
  if (!Number.isInteger(port) || port < 1 || port > 65535) {
    throw new ProxerError("local port must be between 1 and 65535");
  }

  return port;
};

const parseHttpRoute = (input: string | undefined): HttpRouteRequest => {
  if (input === undefined) {
    return { type: "auto" };
  }

  if (input.trim() === "@") {
    return { type: "root" };
  }

  return { type: "subdomain", subdomain: parseHttpSubdomain(input) };
};

export const buildHttpCommand = () => {
  return buildCommand<HttpFlags, [number], ProxerCommandContext>({
    docs: {
      brief: "Expose a local HTTP service through a tunnel.",
    },
    parameters: {
      flags: {
        basicAuthPassword: basicAuthPasswordFlag,
        basicAuthUsername: basicAuthUsernameFlag,
        server: httpServerFlag,
        subdomain: httpSubdomainFlag,
        token: tokenFlag,
      },
      positional: {
        kind: "tuple",
        parameters: [
          {
            parse: parseLocalPort,
            brief: "Local HTTP port to expose.",
            placeholder: "port",
          },
        ],
      },
    },
    async func(flags, localPort) {
      const env = this.env ?? {};
      const token = preferFlag(
        flags.token,
        readEnvString({ env, name: "PROXER_TOKEN" }),
      );
      if (!token) {
        throw new ProxerError("token is required");
      }

      const server =
        preferFlag(
          flags.server,
          readEnvString({ env, name: "PROXER_SERVER" }),
        ) ?? DEFAULT_HTTP_SERVER_URL;
      const envSubdomain = readEnvString({ env, name: "PROXER_SUBDOMAIN" });
      const route = parseHttpRoute(preferFlag(flags.subdomain, envSubdomain));
      const basicAuthPassword = preferFlag(
        flags.basicAuthPassword,
        readEnvString({ env, name: "PROXER_BASIC_AUTH_PASSWORD" }),
      );
      const basicAuthUsername = preferFlag(
        flags.basicAuthUsername,
        readEnvString({ env, name: "PROXER_BASIC_AUTH_USERNAME" }),
      );
      if (basicAuthUsername && !basicAuthPassword) {
        throw new ProxerError(
          "basic auth password is required when username is set",
        );
      }

      await runHttpClient(
        {
          ...(basicAuthPassword
            ? {
                basicAuth: {
                  password: basicAuthPassword,
                  ...(basicAuthUsername ? { username: basicAuthUsername } : {}),
                },
              }
            : {}),
          localPort,
          route,
          serverUrl: resolveControlServerUrl({ server }),
          token,
        },
        { logger: this.logger, process: this.process },
      );
    },
  });
};

File: packages/cli/lib/src/cli/http.dart
import 'package:cliweave/cliweave.dart';
import 'package:proxer/src/cli/command_logger.dart';
import 'package:proxer/src/cli/env.dart';
import 'package:proxer/src/cli/run.dart';
import 'package:proxer/src/cli/shared_flags.dart';
import 'package:proxer/src/config/constants.dart';
import 'package:proxer/src/services/http_client.dart';
import 'package:proxer/src/services/protocol/control_url.dart';
import 'package:proxer/src/services/protocol/frame.dart';
import 'package:proxer/src/services/protocol/subdomain.dart';
import 'package:proxer/src/util/error.dart';

int parseLocalPort(String input) {
  if (!RegExp(r'^\d+$').hasMatch(input)) {
    throw ProxerError('local port must be a number');
  }
  final port = int.parse(input);
  if (port < 1 || port > 65535) {
    throw ProxerError('local port must be between 1 and 65535');
  }
  return port;
}

HttpClientRouteRequest parseHttpRoute(String? input) {
  if (input == null) {
    return const AutoRouteRequest();
  }
  if (input.trim() == '@') {
    return const RootRouteRequest();
  }
  return SubdomainRouteRequest(parseHttpSubdomain(input));
}

final Command httpCommand = buildCommand(
  docs: const CommandDocs(
    brief: 'Expose a local HTTP service through a tunnel.',
  ),
  func: (context, flags, positional) async {
    final logger = loggerFor(context);
    final token = preferFlag(
      flags['token'] as String?,
      readEnvString(context, 'PROXER_TOKEN'),
    );
    if (token == null || token.isEmpty) {
      throw ProxerError('token is required');
    }
    final password = preferFlag(
      flags['basicAuthPassword'] as String?,
      readEnvString(context, 'PROXER_BASIC_AUTH_PASSWORD'),
    );
    final username = preferFlag(
      flags['basicAuthUsername'] as String?,
      readEnvString(context, 'PROXER_BASIC_AUTH_USERNAME'),
    );
    if (username != null && password == null) {
      throw ProxerError('basic auth password is required when username is set');
    }
    final environmentMode = readEnvString(context, 'PROXER_MODE');
    final mode = preferFlag<RouteMode>(
      flags['mode'] as RouteMode?,
      environmentMode == null ? null : parseRouteMode(environmentMode),
    );
    final server =
        preferFlag(
          flags['server'] as String?,
          readEnvString(context, 'PROXER_SERVER'),
        ) ??
        defaultHttpServerUrl;
    await runHttpClient(
      HttpClientConfig(
        localPort: positional.first as int,
        serverUrl: resolveControlServerUrl(server),
        token: token,
        mode: mode ?? RouteMode.single,
        route: parseHttpRoute(
          preferFlag(
            flags['subdomain'] as String?,
            readEnvString(context, 'PROXER_SUBDOMAIN'),
          ),
        ),
        basicAuth: password == null
            ? null
            : BasicAuthConfig(password: password, username: username),
      ),
      info: logger.info,
      error: logger.error,
    );
    return null;
  },
  parameters: const CommandParameters(
    flags: {
      'basicAuthPassword': basicAuthPasswordFlag,
      'basicAuthUsername': basicAuthUsernameFlag,
      'mode': routeModeFlag,
      'server': httpServerFlag,
      'subdomain': httpSubdomainFlag,
      'token': tokenFlag,
    },
    positional: TuplePositionalParameters([
      PositionalParameter(
        brief: 'Local HTTP port to expose.',
        parse: parseLocalPort,
        placeholder: 'port',
      ),
    ]),
  ),
);
